### ISSUE TITLE
Improve speed when building menus

### ISSUE THREAD

**Issue by heatherleaf** [MUSTE-Project/MULLE#23]
It takes an awful lot of time to build the menus. Here are some timings for some examples (in the CLI):

- 1.5s: PrimaSwe, "vinet är vist"
- 2s: PrimaSwe, "ofta håller den stora kejsaren ett gott vin"
- 2s: PrimaLat, "Augustus felix est"
- 11s: ExemplumSwe, "flickan slår på en dator"
- 28s: ExemplumSwe, "flickan i Paris slår på många datorer i dåliga böcker"

The timings are approx similar for the UI, except that it creates menus for two sentences. So about twice the time for the UI, every time a new sentence is selected.

Is it possible to reduce the times by 90-99%? I.e., ca 0.2s for each menu?

(Timings done on my Macbook Pro 2013, so not the quickest computer but not the slowest either)

**Comment by fredefox** [comment]
I've been looking at this. Running a profiler and trying to a bit of digging in the debugger. I've improved performance in some places, but the problem is really the core algorithm which is super inefficient. Fixing this will require a rethink of how we generate adjunction trees. The problems with performance seem to be mainly in the pruning module. I'll have to invest some more time in trying to understand your algorithm to see if I can come up with something clever.

**Comment by heatherleaf** [comment]
Probably the current algorithm can be optimized, but I don't think it will solve the problem completely. The algorithm itself has to be improved -- something for me to think more about.

**Comment by heatherleaf** [comment]
[This comment is also copied from issue [#24](https://github.com/MUSTE-Project/MULLE/issues/24#issuecomment-420639405)]

We currently use the following, requiring that `pruned` and `pruned'` (where `pruned'` comes from the 100,000 adj.trees) have exactly the same metas:

https://github.com/MUSTE-Project/MULLE/blob/2a9b67811d53b4817bb0003bb430fbb59edd460b/muste/src/Muste/Prune.hs#L213-L216

The metas in an adj.tree is static information, so it should be possible to optimise `AdjunctionTrees` quite a bit. Now it's a `Map Category [TTree]`, where the `TTree`s are the adj.trees for the given category.

```
-- belongs to `Muste.AdjunctionTrees.Internal`
-- invariant: and [ c == topNode t && length m == length (filter isMeta (flatten t)) | 
--                  (c,innermap) <- toList adjtrees, (m,ts) <- toList innermap, t <- ts ]
newtype AdjunctionTrees
      = AdjunctionTrees (Map Category (Map Metas [TTree]))

 -- invariant: this should rather be a bag, which we can implement as an ordered list
type Metas = [Category]

-- invariant: length metas == length branches
-- invariant: length (filter isMeta (flatten pruned)) == length metas
pruneTree :: TTree -> [(TTree, Metas, [TTree])]
pruneTree tree = [ (pruned, metas, branches) | ... ]

simTrees :: Map Metas [TTree] -> TTree -> [SimTree]
simTrees adjTreesForCat tree = do
  (pruned, metas, branches) <- pruneTree tree
  pruned' <- filterTrees (adjTreesForCat `lookup` metas) pruned
  tree'   <- insertBranches branches pruned'
  pure (tree `treeDiff` tree', tree', pruned, pruned')
```

Now `guardHeuristics` can get rid of the choices `PRUNE_ALT_2A` and `PRUNE_ALT_2B`

`Muste.AdjunctionTrees.getAdjunctionTrees` also has to be updated of course, which is nontrivial. Currently the internal go function has type

    go :: Category -> (Category, [TTree])

which has to change to

    go :: Category -> (Category, Map Metas [TTree])

Are my ideas clear enough?

**Comment by fredefox** [comment]
I see what you're getting at. Very valuable input. For some reason I missed the comment in #24.

**Comment by fredefox** [comment]
I think I'll start with the change from `[Category]` to `Bag Category`, because that's a rather simple change, and from the profiler I can see that this can have a big impact as well.

**Comment by fredefox** [comment]
Bags landed in 652ab8564c72a92c85c2ed95ec565ff03397b702. Can I change `Grammar.getFunctions` to use it as well?

**Comment by heatherleaf** [comment]
In what way do you want to change getFunctions?

**Comment by fredefox** [comment]
Like so: 5306a1f

**Comment by fredefox** [comment]
Ahhh. This is way better. Thank you for pointing out a good place to improve upon performance. The change was rather trivial. Only `getAdjunctionTrees` looks a bit messy now.

**Comment by fredefox** [comment]
This brings down the time spent getting functions (`Muste.Grammar.getFunctions`) and getting metas (`Muste.Grammar.getMetas`) to half of what it was.

**Comment by fredefox** [comment]
A load time of 1.41s (Exemplum Pars). for a web-site is still on the heavy side for a web site, but perhaps it's more acceptable in this case given that it's not your average run-of-the-mill web site. What should we say the status of this ticket is for now?

**Comment by fredefox** [comment]
Uh! A regression has been introduced, likely related to this work. If you try to open Secunda Pars the server will throw:

    TerminateSessionException Prelude.foldl1: empty list

Do not use `foldl1` and friends! Program defensively or be explicit about what cases you think are unreachable.

**Comment by fredefox** [comment]
The error is thrown in `Muste.Menu.New.collectMenuItems.allNewNodes` at as can be seen with: a849252

**Comment by heatherleaf** [comment]
Good! I would say the status is no longer a "bug", but could probably still be improved.

Perhaps it's time to do some cleanup now? E.g., `PRUNE_ALT_2A` and `PRUNE_ALT_2B` in `heuristics` should be true by default now, so can be removed. 

Also, `heuristics` calls `getFunctions pruned` lot of times. But I guess this info could be part of `AdjunctionTrees`.

**Comment by heatherleaf** [comment]
I don't have much time now until next week, hope you have enough to do until then :)

**Comment by fredefox** [comment]
> Uh! A regression has been introduced, likely related to this work.

This is incorrect. The problem is also in `0.2.4.0`

**Comment by fredefox** [comment]
> I don't have much time now until next week, hope you have enough to do until then :)

Well... Let me see what I can come up with.

- [ ] Improve performance
- [ ] Clean up
  - [ ] Remove CPP pragmas in `Prune`
  - [ ] Make `Muste.Menu.New` the default - i.e. remove the old one.

Improve performance is a non-functional requirement and as such is something that you can do ad infinitum - not sure this is helpful. I think the only other reasonable idea I have is to do some kind of caching. This will make serving requests for the common case seem instantaneous, but I don't know how useful it would be in general.

As for clean up, I think I can accomplish this relatively quickly. I'm a perfectionist/stickler at times so refactoring is also something I could do ad infinitum if given half a chance, but this is of course not desirable.

This just leaves the last two issues we're tracking. I've marked them both as on-hold because they were not a priority, but also because it seems we need a bit more discussion on this before I can get started on it.

TL;DR: I'm not sure I have enough work to keep me occupied for a week.

**Comment by heatherleaf** [comment]
ah.. I meant until beginning of next week :) there's a paper deadline Tuesday..

One kind of profiling is to go in detail into how the menus are created. 

How many adjunction trees are there per key in the map (i.e. per `(Cat, Multiset Cat)` pair)? E.g., for ExemplumSwe, I think the largest is `("Utt", ["NP"])` with >15,000 trees. How does it look for the Secunda grammar?

How large are these trees? (size = `countNodes`) How deep?

What will happen if we limit the maximum size of an adjunction tree? (Or the maximum depth)
- to test: for some (medium- and large) example sentences, how many (and which) menu items will disappear if we insert a maximum size of adjunction trees? 
- we can also limit `Prune.pruneTree` to only returning trees up to a maximum size / depth; how will this affect the final menus?

**Comment by heatherleaf** [comment]
I think the original implementation did that, but then it changed in late spring. The comment didn't change though.

**Comment by fredefox** [comment]
Bah. Just realized something. Why don't we simply cache the `AdjunctionTrees`? There are not that many different `Grammar`s so this can't take up that much space. That should shave off around 20% of the time.

**Comment by fredefox** [comment]
To answer your question about what AdjunctionTrees now looks like. These are the most common combinations:

```
((Cl,	[				]), 2142)
((S,	[				]), 2142)
((Utt,	[				]), 8568)
((Utt,	[(A, 1)				]), 1428)
((Utt,	[(A, 1), (NP, 1)		]), 2170)
((Utt,	[(AP, 1)			]), 1428)
((Utt,	[(AP, 1), (CN, 1), (NP, 1)	]), 1750)
((Utt,	[(AP, 1), (NP, 1)		]), 2170)
((Utt,	[(Adv, 1)			]), 2142)
((Utt,	[(CN, 1), (NP, 1)		]), 1540)
((Utt,	[(CN, 1), (NP, 1), (PN, 1)	]), 1540)
((Utt,	[(Det, 1)			]), 1512)
((Utt,	[(Det, 1), (NP, 1)		]), 2772)
((Utt,	[(N, 1), (NP, 1)		]), 1540)
((Utt,	[(NP, 1)			]), 15876)
((Utt,	[(NP, 1), (Prep, 1)		]), 4284)
((Utt,	[(NP, 1), (V2, 1)		]), 2268)
((Utt,	[(NP, 2)			]), 3829)
((Utt,	[(NP, 2), (Prep, 1)		]), 1155)
((Utt,	[(V2, 1)			]), 1224)
```

**Comment by heatherleaf** [comment]
Caching: yes, good idea. 

N:o of trees: How do these look like for the Prima and Secunda grammars?

**Comment by fredefox** [comment]
I've added a CPP flag enabling diagnostics in 1772b7d76376e5710374a4372cfa01428d537e97. Compile with `-DDIAGNOSTICS` to get at the info. Such flags can be passed to GHC using stack with:

    stack install --ghc-options="-DDIAGNOSTICS"

This will totally flood your terminal with information when you run the web GUI and click a lesson.

**Comment by fredefox** [comment]
[prima.txt](https://github.com/MUSTE-Project/MULLE/files/2382560/prima.txt)
[secunda.txt](https://github.com/MUSTE-Project/MULLE/files/2382561/secunda.txt)

**Comment by fredefox** [comment]
It's my belief that the sizes of the chunks of trees is now more or less negligible. If I modify the code so each chunk only has one tree, then these are the response times:

    exemplum: 646ms
    prima: 770ms

And unrestricted:

    exemplum: 1590ms
    prima: 1540ms

I'm not including secunda because it's throwing.

**Comment by heatherleaf** [comment]
I don't know what you mean with a "chunk" here. But here are some timings on my macbook:

- "en flicka slår på en dator": **0.4 s**
- "en flicka i Paris slår på en dator": **1.1 s**
- "en flicka i Paris slår på en dator i ett hus": **5.2 s**
- "en mor till en flicka i Paris slår på en dator i ett hus": **26 s**

(The times are after the adjunction trees have been pre-calculated. The pre-calculation takes roughly 1 s, which is ok I think. Also, as you mentioned before, that can be stored in a file for quick access.)

So, it seems that the time complexity is roughly exponential in the length of the sentence. It's not an urgent problem, since most of our training sentences will be quite short, but it's something to look into. 

Note: The number of menu items does not increase. They are never more that 30-40 for any selection.

My first suggestion is to limit the tree sizes (both in the adjunction trees and the pruned trees). Introduce a parameter (or two) which we can play around with to see if there's a cutoff where the number of menu items starts to decrease.

**Comment by fredefox** [comment]
I've managed to improve the speed of generating menus. On
`0.2.4.0` (before I started work on this) we had:

    muste-cli "flickan i Paris slår på många datorer i dåliga böcker"  15.91s user 0.10s system 99% cpu 16.015 total

Now we have

    muste-cli "flickan i Paris slår på många datorer i dåliga böcker"  3.33s user 0.06s system 99% cpu 3.396 total

I.e. we went from 15.91s to 3.33s -- a saving of around
80%. This is still not quite the bench-mark that you were hoping
for. You mentioned .2s. I don't have immediate ideas on how to
lower this further.

The bench-mark above was done with the linux `time`
command and therefore includes time for parsing the grammar
file. To get around this I wrote a benchmarking suite that could
give a more accurate picture.  Unfortunately these changes
requires some upstream changes to `pgf` that I am doubtful will be
merged in.  Specifically they require an instance for `NFData`
for `PGF` (this is needed to fully evaluate the grammar before
starting the benchmark). I tried implementing this as orphan
instances but was unable to. I could invest more time into this
and see if it is at all possible to do this in a way that does
not require changes upstream.

The output from the benchmarking suite is included here:


    benchmarking Menu.getMenuItems/vinet är vist
    time                 123.5 μs   (122.3 μs .. 124.8 μs)
                         0.999 R²   (0.999 R² .. 1.000 R²)
    mean                 124.1 μs   (123.4 μs .. 125.2 μs)
    std dev              3.018 μs   (2.032 μs .. 4.170 μs)
    variance introduced by outliers: 19% (moderately inflated)

    benchmarking Menu.getMenuItems/ofta håller den stora kejsaren ett gott vin
    time                 124.2 μs   (123.4 μs .. 125.0 μs)
                         1.000 R²   (0.999 R² .. 1.000 R²)
    mean                 124.5 μs   (123.9 μs .. 125.6 μs)
    std dev              2.623 μs   (1.651 μs .. 4.359 μs)
    variance introduced by outliers: 15% (moderately inflated)

    benchmarking Menu.getMenuItems/Augustus felix est
    time                 123.3 μs   (122.5 μs .. 124.2 μs)
                         1.000 R²   (0.999 R² .. 1.000 R²)
    mean                 123.9 μs   (123.4 μs .. 124.6 μs)
    std dev              2.003 μs   (1.480 μs .. 2.783 μs)
    variance introduced by outliers: 10% (moderately inflated)

    benchmarking Menu.getMenuItems/flickan slår på en dator
    time                 194.3 ms   (188.0 ms .. 200.0 ms)
                         0.999 R²   (0.997 R² .. 1.000 R²)
    mean                 192.7 ms   (188.5 ms .. 195.5 ms)
    std dev              4.757 ms   (2.280 ms .. 6.726 ms)
    variance introduced by outliers: 14% (moderately inflated)

    benchmarking Menu.getMenuItems/flickan i Paris slår på många datorer i dåliga böcker
    time                 2.758 s    (2.712 s .. 2.809 s)
                         1.000 R²   (1.000 R² .. 1.000 R²)
    mean                 2.720 s    (2.700 s .. 2.739 s)
    std dev              21.90 ms   (18.20 ms .. 24.08 ms)
    variance introduced by outliers: 19% (moderately inflated)

As you can see the performance for the last sentence is still pretty bad.

**Comment by fredefox** [comment]
I found out that criterion (that I used for benchmarking) has some other nice goodies included. For instance it can generate these nice looking reports. As you can see from the a lot of variance is introduced by outliers. This is likely due to the issue mentioned above.

The attachment is an HTML, but GitHub will not accept files with a `.html`-suffix -.-

[benchmark.html.txt](https://github.com/MUSTE-Project/MULLE/files/2396023/benchmark.html.txt)

What should I start work on now?

**Comment by heatherleaf** [comment]
Nice graph! 

There are still some smaller cleaning and refactoring that can be done. E.g.:

- move `Menu.Internal.similarTrees` to `Prune`
- get rid of `Prune.replaceTrees` (it's returning a Map which `similarTrees` directly removes)
- probably some more in the web interface etc. (but I don't know that)

Slightly larger is to introduce a size limit of adjunction trees. Meaning:

- `getAdjunctionTrees` should have an extra (Int) argument giving a maximum tree size
- `pruneTree` should also have an (Int) argument with the maximum size of the pruned tree (i.e. the size of `t` in the result pair `(t,bs)` in line 291)

Of course, best is if the too large trees are filtered already when they are build, not in a separate filter phase afterwards.

**Comment by heatherleaf** [comment]
After that I would like to have the option to call `muste-cli` with a max adjunction/pruned tree size, so that I can test if the results degrade with certain values.

**Comment by fredefox** [comment]
`similarTrees` is used within `Muste.Menu.Internal`:

    similarTrees <- collectTreeSubstitutions <- getMenuItems <- getMenu

**Comment by heatherleaf** [comment]
Yes I know, but I just meant to move the definition to `Prune`, and to "inline" `replaceTrees` inside it, because it seems that first `replaceTrees` creates a map which `similarTrees` directly destroys.

**Comment by fredefox** [comment]
I've changed the signature of `27d282993aef75e6e4bc6e6c59e7f5c2a91ae3d0`. Do you think this fixes it? If we want to move `similarTrees` we need to move it's dependents as well.

**Comment by heatherleaf** [comment]
Yes, that also works fine!

**Comment by heatherleaf** [comment]
Closing this now, the remaining issue is now in #28
